<a href="https://colab.research.google.com/github/JoshuneArriaga/Apendizaje_Automatico/blob/main/Tarea_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [5]:
df = pd.read_csv("/content/newborn_health_monitoring_with_risk.csv")
df

baby_id   name  gender  gestational_age_weeks  birth_weight_kg  \
0       B001   Sara  Female                   40.2             3.30   
1       B001   Sara  Female                   40.2             3.30   
2       B001   Sara  Female                   40.2             3.30   
3       B001   Sara  Female                   40.2             3.30   
4       B001   Sara  Female                   40.2             3.30   
...      ...    ...     ...                    ...              ...   
2995    B100  Brian    Male                   39.1             4.47   
2996    B100  Brian    Male                   39.1             4.47   
2997    B100  Brian    Male                   39.1             4.47   
2998    B100  Brian    Male                   39.1             4.47   
2999    B100  Brian    Male                   39.1             4.47   

      birth_length_cm  birth_head_circumference_cm        date  age_days  \
0                50.0                         31.9  2024-12-10         1   
1                50.0                         31.9  2024-12-11         2   
2                50.0                         31.9  2024-12-12         3   
3                50.0                         31.9  2024-12-13         4   
4                50.0                         31.9  2024-12-14         5   
...               ...                          ...         ...       ...   
2995             50.8                         36.3  2025-03-31        26   
2996             50.8                         36.3  2025-04-01        27   
2997             50.8                         36.3  2025-04-02        28   
2998             50.8                         36.3  2025-04-03        29   
2999             50.8                         36.3  2025-04-04        30   

      weight_kg  ...  oxygen_saturation   feeding_type  \
0          3.28  ...                 98        Formula   
1          3.35  ...                 99        Formula   
2          3.35  ...                 97  Breastfeeding   
3          3.48  ...                 97          Mixed   
4          3.50  ...                 96  Breastfeeding   
...         ...  ...                ...            ...   
2995       5.29  ...                 96        Formula   
2996       5.23  ...                 98  Breastfeeding   
2997       5.29  ...                 98  Breastfeeding   
2998       5.39  ...                 97          Mixed   
2999       5.41  ...                 97          Mixed   

      feeding_frequency_per_day  urine_output_count  stool_count  \
0                            11                   9            0   
1                            11                   5            2   
2                             7                   6            5   
3                            11                   8            4   
4                            11                   7            1   
...                         ...                 ...          ...   
2995                          8                   7            4   
2996                          7                   6            1   
2997                          8                   4            1   
2998                         10                   5            2   
2999                          9                   7            1   

      jaundice_level_mg_dl apgar_score  immunizations_done  reflexes_normal  \
0                      4.2         7.0                 Yes              Yes   
1                     10.6         NaN                  No               No   
2                      9.8         NaN                  No               No   
3                      8.1         NaN                  No              Yes   
4                     11.6         NaN                  No              Yes   
...                    ...         ...                 ...              ...   
2995                   4.0         NaN                  No              Yes   
2996                   2.4         NaN                  No              Yes   
2997          

In [6]:
df['apgar_score'] = df.groupby('baby_id')['apgar_score'].ffill()
df['apgar_score'] = df.groupby('baby_id')['apgar_score'].bfill()

In [7]:
numericas = [
    'gestational_age_weeks',
    'birth_weight_kg',
    'birth_length_cm',
    'birth_head_circumference_cm',
    'age_days',
    'weight_kg',
    'length_cm',
    'head_circumference_cm',
    'temperature_c',
    'heart_rate_bpm',
    'respiratory_rate_bpm',
    'oxygen_saturation',
    'feeding_frequency_per_day',
    'urine_output_count',
    'stool_count',
    'jaundice_level_mg_dl',
    'apgar_score'
]
x = df[numericas].copy()
x

gestational_age_weeks  birth_weight_kg  birth_length_cm  \
0                      40.2             3.30             50.0   
1                      40.2             3.30             50.0   
2                      40.2             3.30             50.0   
3                      40.2             3.30             50.0   
4                      40.2             3.30             50.0   
...                     ...              ...              ...   
2995                   39.1             4.47             50.8   
2996                   39.1             4.47             50.8   
2997                   39.1             4.47             50.8   
2998                   39.1             4.47             50.8   
2999                   39.1             4.47             50.8   

      birth_head_circumference_cm  age_days  weight_kg  length_cm  \
0                            31.9         1       3.28       50.1   
1                            31.9         2       3.35       50.2   
2                            31.9         3       3.35       50.6   
3                            31.9         4       3.48       50.3   
4                            31.9         5       3.50       50.3   
...                           ...       ...        ...        ...   
2995                         36.3        26       5.29       52.4   
2996                         36.3        27       5.23       52.1   
2997                         36.3        28       5.29       52.3   
2998                         36.3        29       5.39       52.4   
2999                         36.3        30       5.41       52.3   

      head_circumference_cm  temperature_c  heart_rate_bpm  \
0                      32.0           37.0             160   
1                      32.0           37.2             156   
2                      32.1           37.1             137   
3                      32.0           37.4             130   
4                      31.8           36.8             138   
...                     ...            ...             ...   
2995                   36.9           36.7             125   
2996                   36.8           37.0             138   
2997                   36.7           36.8             143   
2998                   37.0           37.4             119   
2999                   36.9           37.0             144   

      respiratory_rate_bpm  oxygen_saturation  feeding_frequency_per_day  \
0                       36                 98                         11   
1                       40                 99                         11   
2                       37                 97                          7   
3                       33                 97                         11   
4                       39                 96                         11   
...                    ...                ...                        ...   
2995                    44                 96                          8   
2996                    31                 98                          7   
2997                    42                 98                          8   
2998                    52                 97                         10   
2999                    40                 97                          9   

      urine_output_count  stool_count  jaundice_level_mg_dl  apgar_score  
0                      9            0                   4.2          7.0  
1                      5            2                  10.6          7.0  
2                      6            5                   9.8          7.0  
3                      8            4                   8.1          7.0  
4                      7            1                  11.6          7.0  
...                  ...          ...                   ...          ...  
2995                   7            4                   4.0         10.0  
2996                   6            1                   2.4         10.0  
2997                   4            1                   3.6    

In [8]:
#Variable objetivo (predecir peso)
y = x['weight_kg']
X = x.drop(columns=['weight_kg'])

#Dividir datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [9]:
modelo = LinearRegression()
modelo.fit(X_train, y_train)

#Predicciones
y_pred = modelo.predict(X_test)

In [10]:
#Evaluar desempeño
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

In [11]:
print("MÉTRICA")
print(f"MAE: {mae:.4f}")
print(f"MSE: {mse:.4f}")
print(f"RMSE: {rmse:.4f}")
print(f"R²: {r2:.4f}")

MÉTRICA
MAE: 0.0416
MSE: 0.0026
RMSE: 0.0513
R²: 0.9935


Se aplicó un modelo de regresión lineal múltiple para predecir el peso actual del recién nacido a partir de variables  como la edad gestacional, talla, perímetro, signos vitales y frecuencia de alimentación. Los resultados del modelo reflejan que hay una alta precisión por error promedio menor a 0.05 kg y un ajuste casi muy alto, lo que indica que el 99.35 % de la variabilidad del peso del recién nacido es explicada por las variables predictoras incluidas.